pkg/pillar/cmd/msrv/activatecred.go

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) 2023 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package msrv
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+
+	"github.com/google/go-tpm/legacy/tpm2"
+	"github.com/google/go-tpm/tpmutil"
+	etpm "github.com/lf-edge/eve/pkg/pillar/evetpm"
+)
+
+// ActivateCredTpmParam provides the parameters required to activate the
+// TPM credential using EK and AIK.
+type ActivateCredTpmParam struct {
+	Ek      string `json:"ek"`
+	AikPub  string `json:"aikpub"`
+	AikName string `json:"aikname"`
+}
+
+// ActivateCredGenerated contains the generated credential and data to sign.
+type ActivateCredGenerated struct {
+	Data   string `json:"data"`
+	Cred   string `json:"cred"`
+	Secret string `json:"secret"`
+}
+
+// ActivateCredActivated contains the activated credential (decrypted secret)
+// and signature of the ActivateCredGenerated.Data.
+type ActivateCredActivated struct {
+	Secret string `json:"secret"`
+	Sig    string `json:"sig"`
+}
+
+// handles the GET request /tmp/activate-credential/, this is used to get the
+// TPM's EK public key, AIK public key and AIK name.
+func getActivateCredentialParams() ([]byte, []byte, []byte, error) {
+	rw, err := tpm2.OpenTPM(etpm.TpmDevicePath)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	defer rw.Close()
+
+	ekPub, _, _, err := tpm2.ReadPublic(rw, etpm.TpmEKHdl)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	ekPubByte, err := ekPub.Encode()
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	var aikName tpmutil.U16Bytes
+	aikPub, aikName, _, err := tpm2.ReadPublic(rw, etpm.TpmAIKHdl)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	aikPubByte, err := aikPub.Encode()
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	aikNameMarshaled := &bytes.Buffer{}
+	if err := aikName.TPMMarshal(aikNameMarshaled); err != nil {
+		return nil, nil, nil, err
+	}
+
+	return ekPubByte, aikPubByte, aikNameMarshaled.Bytes(), nil
+}
+
+// handles the POST request /tmp/activate-credential/, this is used to activate
+// the credential (decrypt the secret) using EK and AIK.
+func activateCredential(jsonData []byte) ([]byte, []byte, error) {
+	rw, err := tpm2.OpenTPM(etpm.TpmDevicePath)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer rw.Close()
+
+	var credPayload ActivateCredGenerated
+	if err := json.Unmarshal(jsonData, &credPayload); err != nil {
+		return nil, nil, err
+	}
+
+	credBlob, err := base64.StdEncoding.DecodeString(credPayload.Cred)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	encryptedSecret, err := base64.StdEncoding.DecodeString(credPayload.Secret)
+	if err != nil {
+		return nil, nil, err
```

Comment:
```suggestion
		return nil, nil, fmt.Errorf("decoding of secret failed: %w", err)
```
otherwise it is hard to see if the error comes from this line or from https://github.com/lf-edge/eve/pull/4132/files#diff-6be1e60f5fd3f15384691ad26772a5f592c9c5124fb20abf0fe1fd2624a284b2R93

Reply:
good point, fixed.